test/__snapshots__/index.js.snap

```diff
@@ -67,6 +67,54 @@ const Test = () => <div data-jsx={1535297024}>
   </div>;"
 `;
 
+exports[`works with array children 1`] = `
+"import _JSXStyle from \\"styled-jsx/style\\";
+/* eslint-disable no-unused-vars */
+
+const Test1 = () => [<div data-jsx={1486255031}>Foo</div>, <span data-jsx={1486255031}>Bar</span>, <_JSXStyle styleId={1486255031} css={\\"div[data-jsx=\\\\\\"1486255031\\\\\\"]{color:red}\\"} />];
+
+const Test2 = () => [<p data-jsx={2639504339}>Hello World</p>, <_JSXStyle styleId={2639504339} css={\\"p{color:white}\\"} />];
+
+const Test3 = () => [<div data-jsx={4193872369}>
+    <p data-jsx={4193872369}>Hello World</p>
+    <_JSXStyle styleId={4193872369} css={\\"p[data-jsx=\\\\\\"4193872369\\\\\\"]{color:blue}\\"} />
+  </div>, <div>Hello World</div>];
+
+const Test4 = () => [<div data-jsx={493673878}>
```

Comment:
This case is the one that's giving me the most trouble.

I'm not sure what the desired behavior would be in this situation but, as you can see, if there is a div element with a nested `<style>` component **and** a top-level `<style>` component, the top level component wins.

IMHO, this situation should merge the nested style rules with the top level style rules in each of the nested style components and keep the style rules as-is in the top-level component. 

Any ideas?

Reply:
yeah the reasons why we only support top level only are

1) a subtree might not be rendered under some circumstances and that will cause styles to unmount for all the instances e.g.:

```jsx
<div>
   {this.state.isOpen && 
       <div>
          open
          <style jsx>{`div { color: red }`}</style>
       </div>
   }
</div>
```

This will be solved by dynamic styles and another feature I have in mind that will add conditions to tags using this syntax `<style jsx when={expression}>{...}</style>`, in the example `expression` would be `this.state.isOpen`.

2) a problem with nested tags is that the user might think that the styles apply only to the closest scope. In the example above to the closest `div` only not the outer one (the root).

We can try to find a solution that makes the most sense.

Reply:
I see.

What about literally propagating the actual style text down the tree, rather than trying to do fancy stuff with the scoped selector?

**Example:** 

Given this component tree...

```js
export default () => [
  <div>
    {this.state.isOpen && (
      <div>
        open
        <style jsx>{`
          div {
            color: red;
          }
        `}</style>
      </div>
    )}
    <style>{`div { color: blue; }`}</style>
  </div>,
  <style jsx>{`
    div {
      color: white;
    }
  `}</style>
];

```

The output would render like this....

```js
export default () => [
  <div>
    {this.state.isOpen && (
      <div>
        open
        <style jsx>{`
          div {
            color: white;
          }
          div {
            color: blue;
          }
          div {
            color: red;
          }
        `}</style>
      </div>
    )}
    <style>{`
	div { color: white; }
	div { color: blue; }
   `}</style>
  </div>,
  <style jsx>{`
    div {
      color: white;
    }
  `}</style>
];
```

That way, any duplicate selectors would automatically take precedence in the child `<style>` blocks and any non-dupes would still propagate downward?

Thoughts? Am I totally leaving out something obvious that makes this a more challenging?

Reply:
uhm I am not sure about that. In my opinion styles like that should apply as if they were contained inside of the root element without us having to do that kind of "propagation".

I haven't tried this yet but I think that if you don't decrease `state.ignoreClosing` here https://github.com/zeit/styled-jsx/pull/284/files#diff-bb35c33015964b91f5c83deb2a33a814R263 it might work. Basically for now we shouldn't worry about nested styles, we just want the following to work:


```jsx
() => [
  <div>
      hi
      <style jsx>{`div { font-weight: bold }`}</style>
  </div>,
  <p>there</p>,
  <style jsx>{`div { color: red }`}</style>
]
```

Reply:
or just warn / error when `path.parentNode` of a style tag is an `ArrayExpression` :)

Reply:
You're right. Just tried to quickly write out what the process would look like if we were to propagate styles downward and it quickly became overly complex with no real benefit.

Question on your proposition:

In your example above, what would your output be?

You have a single div with conflicting style rules, one from a child style element and one from the root style element. 

If the logic proceeds as it currently does, only one of those rules will apply to the div in your example (whichever is located first in state.styles, since styles are received by using `Array.shift()`)

Consider the following situation...

```js
() => [
  <div>
    // contents....
    <style>{/* style here */}</style>
  </div>,
  <div>
    // contents....
    <style>{/* style here */}</style>
  </div>,
  <style>{/* style here */}</style>,
];
```

What would be expected to happen here?

Reply:
(Having said that, I'm totally fine with throwing the error as you mentioned. Just making sure we're on the same page.)